src/math_parser_diag.cpp

```diff
@@ -565,6 +565,75 @@ std::function<double( dialogue & )> mod_order_eval( char /* scope */,
     };
 }
 
+enum class character_filter : int {
+    friends = 0,
+    not_friends,
+    both,
+};
```

Comment:
Shouldn't there be an `enemy` attitude too?

Reply:
Probably? I don't understand vectors at all, and there's no function in Character that I'm aware of that works as easily as is_ally() does. There's stuff like guaranteed_hostile() and is_enemy, but that's all locked to the avatar and wouldn't work for NPCs.

Reply:
There's a `Character::attitude_to()` that returns hostile/neutral/friendly. The `gets angry` message is printed when attitude switches to hostile.

Reply:
Right, what I mean is that a check for enemies would fail to count enemies which were 100% hostile but simply had not seen the character yet. See #71609

Reply:
Try something like this
```c++
        case character_filter::hostile:
            return guy.attitude_to( beta ) == Creature::Attitude::HOSTILE ||
                ( beta.is_avatar() && guy.is_npc() && guy.as_npc()->guaranteed_hostile() );
```

Even if that doesn't work, this PR only uses the friendly attitude so a bug with hostile can be fixed later. That'd be easier and require less churn that renaming `friends_nearby()` later in code, documentation, and JSON.

Please either do that or remove the attitude bits that are effectively dead code (`_friend_match_filter_character()` and `character_filter`)

Reply:
Implemented your changes and switched the name of the thing over to characters_nearby. It can now filter by allies, not_allies, hostile, or any. I wanted to include an optional bool to allow halllucinations, but I couldn't figure out how to pass a non-kwarg bool to the function and it's not all that important right now.

I see something in the documentation where under radius the example on monsters_nearby uses u_search_radius * 3, but I cannot find anything anywhere in the code or documentation explaining that. Do you happen to know what it refers to? I've currently got the social/asocial traits just looking for allies ~30 tiles out and that's fine, but if there's an easy way to limit the radius to visual distance or something I'm all ears.

Reply:
> Implemented

Nice; thanks a lot!

> I see something in the documentation where under radius the example on monsters_nearby uses u_search_radius * 3, but I cannot find anything anywhere in the code or documentation explaining that. Do you happen to know what it refers to?

It's just an example for use of a variable in that kwarg. `u_search_radius` is the [variable](https://github.com/CleverRaven/Cataclysm-DDA/blob/master/doc/NPCs.md#variables) `search_radius` stored on the alpha talker.

> I've currently got the social/asocial traits just looking for allies ~30 tiles out and that's fine, but if there's an easy way to limit the radius to visual distance or something I'm all ears.

Vision range is not currently exposed to math, but it should be easy to add.

<details>
<summary>Something like this</summary>

```diff
diff --git a/src/math_parser_diag.cpp b/src/math_parser_diag.cpp
index 8c13059baa..92b106ed90 100644
--- a/src/math_parser_diag.cpp
+++ b/src/math_parser_diag.cpp
@@ -1143,6 +1143,19 @@ std::function<double( dialogue & )> value_or_eval( char /* scope */,
     };
 }
 
+std::function<double( dialogue & )> vision_range_eval( char scope,
+        std::vector<diag_value> const &/* params */, diag_kwargs const &/* kwargs */ )
+{
+    return[beta = is_beta( scope )]( dialogue const & d ) {
+        if( d.actor( beta )->get_character() ) {
+            return static_cast<talker const *>( d.actor( beta ) )
+                   ->get_character()
+                   ->unimpaired_range();
+        }
+        return 0;
+    };
+}
+
 std::function<double( dialogue & )> vitamin_eval( char scope,
         std::vector<diag_value> const &params, diag_kwargs const &/* kwargs */ )
 {
@@ -1290,6 +1303,7 @@ std::map<std::string_view, dialogue_func_eval> const dialogue_eval_f{
     { "val", { "un", 1, u_val } },
     { "value_or", { "g", 2, value_or_eval } },
     { "vitamin", { "un", 1, vitamin_eval } },
+    { "vision_range", { "un", 0, vision_range_eval } },
     { "warmth", { "un", 1, warmth_eval } },
     { "weather", { "g", 1, weather_eval } },
 };
```

I hacked this together quickly and didn't test it.


</details>

Then you can use it from an EOC like any other function, for example
```JSON
    "condition": { "math": [ "u_characters_nearby( 'radius': u_vision_range() ) > 0" ] },
```

Reply:
Alright, I need to test some of this but it's all pushed now. Thanks again for the help.